---
title: 'GSE19711'
output: html_document
---

# Authors


- Rimvydas Noreika
- Tautminas Cibulskis
- Gabrielius Erignis
- Rolandas PorÄ—jus


# Analysis

```{r, echo=FALSE, include=FALSE}
library(GEOquery)
```

-Automatically download the data from GEO
```{r}
gset = getGEO("GSE19711", destdir="./")
```

-Obtain the matrix of beta values where each row corresponds to probes and each column corresponds to samples
```{r}
gsetexpr = exprs(gset[[1]])
rownames(gsetexpr)[1:5]
colnames(gsetexpr)[1:5]
```

-How many samples and how many probes do you have in the data?
```{r}
sampleset = colnames(gsetexpr)
length(sampleset)

probeset = rownames(gsetexpr)
length(probeset)
```

-How are the beta values distributed?
```{r}
 hist(gsetexpr)
```

-Do your probes have names?
```{r}
probeset[1:5]
```

-Do you have annotation that tells the coordinate (in hg19) of each probe and its genomic features (such as related gene name)?
```{r}
annotation = getGEO("GPL8490", destdir = "./")
annotation = Table(annotation)
head(annotation)
```

-Do you know which samples correspond to healthy individuals, and which samples correspond to the sick ones?
```{r}
disease = pData(phenoData(gset[[1]]))
disease[1:20,1]
```
For each probe compute a t-test to verify if the distributions of beta values within the probe significantly differ between the two groups. From the t-test, obtain the p value.
```{r}
pvalarray = array()
namearray = array()
for (i in 1:length(probeset)){
  case = gsetexpr[i, grep("Case", disease$`sample type:ch1`)]
  control = gsetexpr[i, grep("Control", disease$`sample type:ch1`)]
  tt = t.test(case, control)
  pvals <- tt$p.value
  pvalarray[i] = tt$p.value
  namearray[i] = probeset[i] 
}
```

Plot the distribution of p values. What is the expected distribution? How does it differ from what you get?
We get Anti-conservative p-values. All null p-values form a flat distribution. Peak is close to 0 - that is where the alternative hypotheses are. Not all p-values less than .05 are significant. High p-values indicate false negatives - hypotheses that are not detectable with our test. The left of the histogram shows significance. The right side shows how many p-values are null.

```{r}
hist(pvalarray, breaks = 100)
```
Performance-wise, how long will it take to compute the test for all probes?
Result is 51.228 seconds.

```{r}
system.time( for (i in 1:length(probeset)){
  case = gsetexpr[i, grep("Case", disease$`sample type:ch1`)]
  control = gsetexpr[i, grep("Control", disease$`sample type:ch1`)]
  tt = t.test(case, control)
  pvalarray[i] = tt$p.value
})
```
What is multiple hypothesis testing? Such testing means that we are trying to check validity of a number of hypotheses simultaneausly. However, as the number of hypotheses increases, o does the chance for observing at least one significant result by chance. In order to reduce the impact of this isssue, it is important to adjust for multiple hypothesis testing, so observing a significant result by chance is below our chosen significance level.


How should we adjust for multiple hypothesis testing in our case?

-What is multiple hypothesis testing?
Such testing means that we are trying to check validity of a number of hypotheses simultaneausly. However, as the number of hypotheses increases, o does the chance for observing at least one significant result by chance. In order to reduce the impact of this isssue, it is important to adjust for multiple hypothesis testing, so observing a significant result by chance is below our chosen significance level.

-How should we adjust for multiple hypothesis testing in our case?

```{r}
pvalarray.adjusted <- p.adjust(pvalarray, method="BH", n = length(pvalarray))
```
The p-values should be adjusted using "adjust" funtion. It is important to choose the right method to complete this action. In this case, "BH" method is the optimal one, since according to R documentation it controls the false discovery rate (proportion which is expected among the rejected hypotheses). This is exactly what we need in multiple hypothesis testing. It is also mentioned that these methods "are more powerful than others".


Did you find any probes that show statistically significant modification difference between healthy and sick individuals?
```{r}
pvalarray.significant <- pvalarray.adjusted[pvalarray.adjusted < 0.05]
namearray.significant <- namearray[pvalarray.adjusted < 0.05]
namearray.significant [1:20]
```
In this case we define smaller than 0.05 adjusted pvalue as "significant". The variables which are within this criteria are chosen and their names are printed.

Where are these probes? What genes are they related to?

-Did you find any probes that show statistically significant modification difference between healthy and sick individuals?
```{r}
pvalarray.significant <- pvalarray.adjusted[pvalarray.adjusted < 0.04]
namearray.significant <- namearray[pvalarray.adjusted < 0.04]
namearray.significant [1:20]
```
In this case we define smaller than 0.04 adjusted pvalue as "significant". The variables which are within this criteria are chosen and their names are printed. 

-Where are these probes? What genes are they related to?

```{r}
genes <- annotation$Gene_ID[match(namearray.significant, annotation$ID)]
location <- annotation$CPG_ISLAND_LOCATIONS[match(namearray.significant, annotation$ID)]
genes [1:20]
location [1:20]
```
We should find genes and then chromosomes (where probes are located). We do so by matching the names of adjusted pvalues and annotation IDs.


Normalizavimas, kvantiliu normalizacija

```{r}
library(impute)
library(limma)
library(minfi)
normmat = limma::normalizeBetweenArrays(gsetexpr)
```
Some samples could be outliers messing up our results. How to detect outliers?

```{r}
data711 = read.csv('DnaAge.csv')
agesampledata <- data711[,c("DNAmAge","SampleID")]
groups <- agesampledata[2]
dim(normmat)
sum(is.na(normmat))
imputed = impute.knn(normmat)
dim(imputed$data)
sum(is.na(imputed$data))
pca = prcomp(t(imputed$data), scale=FALSE)
head(pca$sdev / sum(pca$sdev)) #variabilumas
x = 1:540
y = (pca$sdev / sum(pca$sdev))
plot(x,y, type='h')
```
PCA

```{r}
d = dist(pca$x[,1:5])
h = hclust(d, method="average")
plot(h)
```

Reikia atmesti lytiniu chromosomu probus

```{r}

```

Permutacijos!!

```{r}
# observed = mean(pvalarray < 0.05)
# set.seed(123)
# expected = numeric(100)
# for(iteration in 1:1000){
#   groupsRandom = sample(groups)
#   pvalsRandom = apply(normmat[1:1000,],1,function(x) {t.test(x ~ groupsRandom)$p.value})
# expected[iteration] = mean(pvalsRandom <0.05)
# print (iteration)
# }
```
Remember that the experiment may have more than one variable (disease). Age may be a factor two. How can we account for differences in age?
```{r}

```
Age and Blood cellular composition may be an issue too. Maybe the differences we observe between patient groups are not actual methylation differences but artefacts of varying blood cellular composition? How shall we deal with that?
```{r}

```
Rezultatas: pakite probai!!!
```{r}
# age <- agesampledata[1]
# model <- model.matrix(~ 1 + unlist(groups) + unlist(age))
# fit <- lmFit(normmat, model)
# fit <- eBayes(fit)
```